changes to enable unified push notifications in generic build. Signed…

gavine99 opened the pull request:
changes to allow a UnifiedPush distributor to deliver push messages in generic/f-droid build.

added a setting item to reset push notification registrations in case of issues with push (UnifiedPush distributor can sometimes get out of sync). this same setting item can also be used to reset firebase push in case of issues with it too.

a similar pr has been submitted for the nextcloud android client app  app - https://github.com/nextcloud/android/pull/13516

this pr is related to, but not dependent on, 3 other pr's;
  nextcloud server - https://github.com/nextcloud/server/pull/47763  (to provide enhanced ability for multiple notification handlers)
  nextcloud notifications app - https://github.com/nextcloud/notifications/pull/2027 (to provide enhanced ability for multiple notification handlers)
  3rd party unified push app - https://codeberg.org/NextPush/uppush/pulls/17 (to allow existing 3rd party nextcloud unifiedpush app to send nextcloud notifications via unifedpush)

p1gp1g commented on app/build.gradle:
```diff
@@ -303,6 +303,9 @@ dependencies {
     implementation 'com.github.nextcloud-deps:android-talk-webrtc:121.6167.0'
 
+    // unified push library for generic flavour
+    genericImplementation 'org.codeberg.UnifiedPush:android-connector:2.4.0'
```
```suggestion
    genericImplementation 'org.unifiedpush.android:connector:2.5.0'
```

Fuseteam commented:
@gavine99 you may want checkout the DCO error message, if you click on the x it will explain what you need to do

gavine99 commented:
@Fuseteam i'm not sure what i'm supposed to do to fix ??  dco results say 8 commits not signed off my 4 commit messages all have sign-off string; `Signed-off-by: Gavin Element <github@bobfox.top>`

the other 4 commits listed dco results are nextcloud android bot and i can't sign them off;

`Commit sha: [36716fa](https://github.com/nextcloud/talk-android/pull/4169/commits/36716fa78871d4bd24376d11adebdb4749d745d1), Author: Nextcloud Android Bot, Committer: Nextcloud Android Bot; The sign-off is missing.
Commit sha: [d450c47](https://github.com/nextcloud/talk-android/pull/4169/commits/d450c470fe56138b1c1bc345d95a67e0adc427d7), Author: Nextcloud Android Bot, Committer: Nextcloud Android Bot; The sign-off is missing.
Commit sha: [be235c8](https://github.com/nextcloud/talk-android/pull/4169/commits/be235c8141e8e1c6906198ace61ae17c307445e6), Author: Nextcloud Android Bot, Committer: Nextcloud Android Bot; The sign-off is missing.
Commit sha: [667a67b](https://github.com/nextcloud/talk-android/pull/4169/commits/667a67b7fc0858d20549133d17386a23a98a6bf1), Author: Nextcloud Android Bot, Committer: Nextcloud Android Bot; The sign-off is missing.`

Fuseteam commented:
@gavine99 perhaps try rebasing your branch and only picking your own commits

Fuseteam commented:
ayyy nice DCO looks good now :D

Fuseteam commented:
looks like remaining issues are listed here: https://app.codacy.com/gh/nextcloud/talk-android/pull-requests/4169/issues

p1gp1g commented:
As I said on [the NextPush PR](https://codeberg.org/NextPush/uppush/pulls/17#issuecomment-2282854), a better approach is to add support for webpush, so we can have a proper UnifiedPush support on the Nextcloud applications: it won't depend on a pre-defined distributor (which goes against the whole UnifiedPush idea)

I will work on adding this support in not so long. Your different PR on the clients may be used as a base for it

gavine99 commented:
> it won't depend on a pre-defined distributor

@p1gp1g maybe i don't understand enough but this pr uses the generic unified push connector so it doesn't depend on a predefined distributor (ie the distributor can use webpush or any other push tech). this pr also uses the same message format as the existing fcm push messages. it's all very generic and doesn't commit anyone to using anything.

gavine99 commented:
> other than for the rebase(s), i believe you shouldn't force push, the maintainer can always squash later. But maybe that's just me

@Fuseteam i rebased and squashed because all my commits had the wrong signed-off-by info

p1gp1g commented:
> > it won't depend on a pre-defined distributor
> 
> @p1gp1g maybe i don't understand enough but this pr uses the generic unified push connector so it doesn't depend on a predefined distributor (ie the distributor can use webpush or any other push tech). this pr also uses the same message format as the existing fcm push messages. it's all very generic and doesn't commit anyone to using anything.

I didn't look at the details, this is what I understood based on your requirements for nextpush. Does it work with another distributor right now ?

gavine99 commented:
there is nothing to make it rely on nextpush. the original pr description does say;  ... this pr is related to, **but not dependent on**,...   i will edit the pr description to make it more clear.

if one has multiple distributors installed (i have nextpush and ntfy) it even allows you to choose which distributor to use using the connector-ui library dialog.

![image](https://github.com/user-attachments/assets/25d4e281-c37a-46b7-aaad-03c9873843cb)

p1gp1g commented:
Well, that's great then. It would be nice to edit the different PR to know what is required, what is not, and what has changed!

If you want, I'm OK to review your different client-side PRs.

From uppush PR:
> allow this app to be used to push nextcloud notifications to nextcloud android client and talk apps.

It isn't true anymore then ? Or it has never been a requirement ?

---

I will still work to add webpush on notifications app, it is needed for different reasons. And it can then be used by the different clients then.

Do you use the same trick than neon at this moment ? It adds a hash `#` to remove the hardcoded path from nextcloud proxy

gavine99 commented:
> > allow this app to be used to push nextcloud notifications to nextcloud android client and talk apps.

> It isn't true anymore then ? Or it has never been a requirement ?

it was never a hard requirement.  that uppush pr can be abandoned because it is no longer required since i have developed the stand-alone nextcloud server app that bridges notifications and uppush. i have been using it for months and it works great.

there are no dependencies for my nextcloud talk and nextcloud app unified push connector pull requests. they are totally independent.

> I will still work to add webpush on notifications app, it is needed for different reasons. And it can then be used by the different clients then.

that sounds good.  do you intend webpush to be used as transport for unified push - ie webpush used between server and unified push distributor?   or totally separate from unified push?

> Do you use the same trick than neon at this moment ? It adds a hash # to remove the hardcoded path from nextcloud proxy

i don't know what neon is, what the hard-coded path is, and what nextcloud proxy you mean  :)  but i can say i didn't do any tricks or work-arounds. i just added the unified push libraries and plumbed them into the existing fcm push receiver code with the least amount of changes to existing code as possible.

Fuseteam commented:
> 
> The specs have been clarified to explicitly define webpush as the server to server protocol. It generally work even if it misses some parts.

if understand correctly does this mean that the nextcloud notifications app will talk to ntfy, uppush or ubports push over webpush?

Fuseteam commented:
then again that nextpush pr is for the uppush the server app— is there is even a nextpush pr for the distributor? looks like i'm still confusing the two :joy:

gavine99 commented:
@Fuseteam @p1gp1g

i'll try to clear things up with the pushes;

**mobile apps**
the talk app and the nextcloud app pr's can be merged any time. they are completely separate and have no dependency on any other pr's. they enable any unified push distributor to be used for delivering push messages to the apps. push messages can be in the same format as the existing fcm messages but without the encryption as it's not required because the notifications are no longer going through nextcloud push proxy or google fcm.

**uppush server app**
the pr's for uppush can be abandoned, because as you say @p1gp1g, they 'tied' notifications into the uppush app. the notification-uppush 'bridge' that i developed (https://github.com/gavine99/uppush-notifications) does the job of the old uppush pr's.

**nextcloud server and the nextcloud server notifications app**
these pr's achieved 1 main thing; to allow the deletion 'event' of notifications to go through the 'manager' in main code base so that all registered notification-managing-apps get a call when a notification is deleted. but these pr's are possibly not implemented well. i intend to revisit them soon to see if the same can be achieved in a different fashion. this doesn't change the fact that the mobile app pr's can be merged ahead of any changes on the server.

i hope this is helpful.

gavine99 commented:
> is there is even a nextpush pr for the distributor

there is a pr for the distributor, but it's _completely separate_ bug fix and has no relationship to nextcloud or uppush

gavine99 commented:
> The specs have been clarified to explicitly define webpush as the server to server protocol

for me, i am not a fan of this idea. i think unified push should be transport agnostic.

a reason why is that, in the future when i had time available, i wanted to develop a unified push distributor based on mqtt back-end so that apps can 'push messages back.' (and other benefits that mqtt provides).   yes, i realise this would be a non-standard extension of unified push but it would be unified push compliant for messages from server to client so ok for 100% unified push compliance. it would just do _extra_ stuff. (maybe name it "unified push**back**" ?? :) )

a nice example of what is good about apps being able to 'push messages back' would be that sip protocol could be 100% performed over this.

gavine99 commented:
> it just how web services talk to push servers

sounds great

p1gp1g commented:
@gavine99 I've sent you a message on matrix

Fuseteam commented:
why was this closed?

gavine99 commented:
i deleted my local fork. the maintainer(s) of this repo should still have a way to access to the pr changes
